File: problem92/problem92.c
#include <stdio.h>

int next(int n) {
	int result = 0;
	while (n > 0) {
		result += (n%10) * (n%10);
		n /= 10;
	}
	
	return result;
}

int main() {
	int buffer[600];
	int count = 0;
	
	buffer[1] = 0;
	buffer[89] = 1;
	
	int i = 0;
	for (i = 1; i < 568; i++) {
		int n = i;
		while (n != 1 && n != 89)
			n = next(n);

		if (n == 1)
			buffer[i] = 0;
		else {
			buffer[i] = 1;
			count += 1;
		}
	}
	printf("generated buffer to 600...\n");
	
	for (i = 568; i < 10000000; i++)
		if (buffer[next(i)] == 1)
			count += 1;
	
	printf("%d\n", count);
	return 0;
	
}

File: problem92/problem92.rb
def next_number(n)
  result = 0
  result, n = result + (n%10)**2, n/10 while n > 0
  
  return result
end

MAX = 10_000_000
count = 0
buffer = Hash.new
buffer[1] = false
buffer[89] = true

1.upto(567) do |i|
  n = i
  while !buffer.key?(n)
    n = next_number(n) 
  end
  
  buffer[i] = buffer[n]
  count += 1 if buffer[i]
end
puts "end generate buffer to 567..."

568.upto(MAX-1) do |i|
  count += 1 if buffer[next_number(i)]
end

puts count